lib/cn/spill.c

```diff
@@ -286,243 +310,225 @@ cn_spill(struct cn_compaction_work *w)
     direct_read_len = w->cw_rp->cn_compact_vblk_ra;
     direct_read_len -= PAGE_SIZE;
 
-    memset(w->cw_outv, 0, w->cw_outc * sizeof(*w->cw_outv));
+    *added = false;
+
+    if (!kvcompact) {
+        if (!sctx->more || key_obj_cmp(&sctx->curr.kobj, &ekobj) > 0) {
+            if (!sctx->pt_set)
+                return 0; /* Nothing to do */
+        }
+    }
 
-    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    err = get_kvset_builder(w, 0, &child);
     if (err)
         return err;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (more)
-        cn_spill_route_get(w->cw_tree, &curr.kobj, &ri);
+    assert(child);
 
-    while (!err && more) {
-        bool new_key = true;
-
-        if (!child) {
-            err = get_kvset_builder(w, output_nodec, &child);
-            if (err)
-                break;
-            assert(child);
+    /* Add ptomb to 'child' if a ptomb context is carried forward from the
+     * previous node spill, i.e., this ptomb spans across multiple children.
+     */
+    if (sctx->more && sctx->pt_set && (!w->cw_drop_tombs || sctx->pt_seq > w->cw_horizon)) {
 
-            /* Add ptomb to 'child' if a ptomb context is carried forward from the
-             * previous node spill, i.e., this ptomb spans across multiple children.
-             */
-            if (pt_set && (!w->cw_drop_tombs || pt_seq > w->cw_horizon)) {
+        err = kvset_builder_add_val(child, &sctx->pt_kobj, HSE_CORE_TOMB_PFX, 0, sctx->pt_seq, 0);
+        if (!err)
+            err = kvset_builder_add_key(child, &sctx->pt_kobj);
 
-                err = kvset_builder_add_val(child, &pt_kobj, HSE_CORE_TOMB_PFX, 0, pt_seq, 0);
-                if (!err)
-                    err = kvset_builder_add_key(child, &pt_kobj);
+        if (err) {
+            kvset_builder_destroy(child);
+            return err;
+        }
 
-                if (err) {
-                    kvset_builder_destroy(child);
-                    break;
-                }
+        w->cw_stats.ms_keys_out++;
+        w->cw_stats.ms_key_bytes_out += key_obj_len(&sctx->pt_kobj);
+    }
 
-                w->cw_stats.ms_keys_out++;
-                w->cw_stats.ms_key_bytes_out += key_obj_len(&pt_kobj);
-            }
-        }
+    new_key = true;
 
-        while (more && (kvcompact || ri.last_node || key_obj_cmp(&curr.kobj, &ri.ekobj) <= 0)) {
+    tstart = perfc_ison(w->cw_pc, PERFC_DI_CNCOMP_VGET) ? 1 : 0;
```

Comment:
I think we should just get rid of this perf counter, or put the related code under #ifndef HSE_BUILD_RELEASE.

Reply:
I vote lose it.